Mathlib/Topology/MetricSpace/PartitionOfUnity.lean

```diff
@@ -0,0 +1,162 @@
+
+variable {ι X : Type _}
+
+namespace EMetric
+
+variable [EMetricSpace X] {K : ι → Set X} {U : ι → Set X}
+
+/-- Let `K : ι → set X` be a locally finitie family of closed sets in an emetric space. Let
+`U : ι → set X` be a family of open sets such that `K i ⊆ U i` for all `i`. Then for any point
+`x : X`, for sufficiently small `r : ℝ≥0∞` and for `y` sufficiently close to `x`, for all `i`, if
+`y ∈ K i`, then `emetric.closed_ball y r ⊆ U i`. -/
+theorem eventually_nhds_zero_forall_closedBall_subset (hK : ∀ i, IsClosed (K i))
+    (hU : ∀ i, IsOpen (U i)) (hKU : ∀ i, K i ⊆ U i) (hfin : LocallyFinite K) (x : X) :
+    ∀ᶠ p : ℝ≥0∞ × X in 𝓝 0 ×ᶠ 𝓝 x, ∀ i, p.2 ∈ K i → closedBall p.2 p.1 ⊆ U i := by
+  suffices ∀ i, x ∈ K i → ∀ᶠ p : ℝ≥0∞ × X in 𝓝 0 ×ᶠ 𝓝 x, closedBall p.2 p.1 ⊆ U i by
+
+    apply mp_mem ((eventually_all_finite (hfin.point_finite x)).2 this)
+      (mp_mem (@tendsto_snd ℝ≥0∞ _ (𝓝 0) _ _ (hfin.interᵢ_compl_mem_nhds hK x)) _)
+    apply univ_mem'
+    rintro ⟨r, y⟩ hxy hyU i hi
+    simp only [mem_interᵢ, mem_compl_iff, not_imp_not, mem_preimage] at hxy
+    exact hyU _ (hxy _ hi)
+  intro i hi
+  rcases nhds_basis_closed_eball.mem_iff.1 ((hU i).mem_nhds <| hKU i hi) with ⟨R, hR₀, hR⟩
+  rcases ENNReal.lt_iff_exists_nnreal_btwn.mp hR₀ with ⟨r, hr₀, hrR⟩
+  filter_upwards [prod_mem_prod (eventually_lt_nhds hr₀)
+      (closedBall_mem_nhds x (tsub_pos_iff_lt.2 hrR))] with p hp z hz
+  apply hR
+  calc
+    edist z x ≤ edist z p.2 + edist p.2 x := edist_triangle _ _ _
+    _ ≤ p.1 + (R - p.1) := (add_le_add hz <| le_trans hp.2 <| tsub_le_tsub_left hp.1.out.le _)
+    _ = R := add_tsub_cancel_of_le (lt_trans (by exact hp.1) hrR).le
+#align emetric.eventually_nhds_zero_forall_closed_ball_subset EMetric.eventually_nhds_zero_forall_closedBall_subset
+
+theorem exists_forall_closedBall_subset_aux₁ (hK : ∀ i, IsClosed (K i)) (hU : ∀ i, IsOpen (U i))
+    (hKU : ∀ i, K i ⊆ U i) (hfin : LocallyFinite K) (x : X) :
+    ∃ r : ℝ,
+      ∀ᶠ y in 𝓝 x,
+        r ∈ Ioi (0 : ℝ) ∩ ENNReal.ofReal ⁻¹' ⋂ (i) (hi : y ∈ K i), { r | closedBall y r ⊆ U i } :=
+  by
+  have :=
+    (ENNReal.continuous_ofReal.tendsto' 0 0 ENNReal.ofReal_zero).eventually
+      (eventually_nhds_zero_forall_closedBall_subset hK hU hKU hfin x).curry
+  rcases this.exists_gt with ⟨r, hr0, hr⟩
+  refine' ⟨r, hr.mono fun y hy => ⟨hr0, _⟩⟩
+  rwa [mem_preimage, mem_interᵢ₂]
+#align emetric.exists_forall_closed_ball_subset_aux₁ EMetric.exists_forall_closedBall_subset_aux₁
+
+theorem exists_forall_closedBall_subset_aux₂ (y : X) :
+    Convex ℝ
+      (Ioi (0 : ℝ) ∩ ENNReal.ofReal ⁻¹' ⋂ (i) (hi : y ∈ K i), { r | closedBall y r ⊆ U i }) :=
+  (convex_Ioi _).inter <|
+    OrdConnected.convex <|
+      OrdConnected.preimage_ennreal_ofReal <|
+        ordConnected_interᵢ fun i =>
+          ordConnected_interᵢ fun (_ : y ∈ K i) => ordConnected_setOf_closedBall_subset y (U i)
+#align emetric.exists_forall_closed_ball_subset_aux₂ EMetric.exists_forall_closedBall_subset_aux₂
```

Comment:
For both of these functions, the linter says `(hi : y ∈ K i)` is an unused variable, but I'm not sure how to fix it. Changing to `(y ∈ K i)` doesn't work.

Reply:
try `(_hi : y ∈ K i)`